pkg/southbound/synchronizer/device_update.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"errors"
+	"strconv"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/events"
+
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+func (s *Session) getTermPerDevice(device *topodevice.Device) (int, error) {
+	term := device.Attributes[mastershipTermKey]
+	if term == "" {
+		return 0, nil
+	}
+	return strconv.Atoi(term)
+
+}
+
+func (s *Session) updateDevice(connectivity topodevice.ConnectivityState, channel topodevice.ChannelState,
+	service topodevice.ServiceState) error {
+	log.Info("Update device state")
+
+	id := s.device.ID
+	topoDevice, err := s.deviceStore.Get(id)
+	st, ok := status.FromError(err)
+
+	// If the device doesn't exist then we should not update its state
+	if ok && err != nil && st.Code() == codes.NotFound {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	protocolState, index := containsGnmi(topoDevice.Protocols)
+	if protocolState != nil {
+		topoDevice.Protocols = remove(topoDevice.Protocols, index)
+	} else {
+		protocolState = new(topodevice.ProtocolState)
+	}
+
+	protocolState.Protocol = topodevice.Protocol_GNMI
+	protocolState.ConnectivityState = connectivity
+	protocolState.ChannelState = channel
+	protocolState.ServiceState = service
+	topoDevice.Protocols = append(topoDevice.Protocols, protocolState)
+	mastershipState, err := s.mastershipStore.GetMastership(topoDevice.ID)
+	if err != nil {
+		return err
+	}
+
+	// Read the current term for the given device
+	currentTerm, err := s.getTermPerDevice(topoDevice)
+	if err != nil {
+		return err
+	}
+
+	// Do not update the state of a device if the node encounters a mastership term greater than its own
+	if uint64(mastershipState.Term) < uint64(currentTerm) {
+		return errors.New("device mastership term is greater than node mastership term")
```

Comment:
When a higher mastership term is discovered, this node needs to stop attempting this update. But since this function is being called by the `backoff` library, this error will just cause it to retry and create a deadlock since the update will always fail after this condition is met. The error returned has to be `backoff.NewPermanent` or whatever the error was that stops the retries.

Reply:
Ooops. You are right. We should not return the error here.